packages/replayio/src/utils/createPromiseQueue.ts

```diff
@@ -0,0 +1,91 @@
+import { Deferred, createDeferred } from "./createDeferred";
+
+interface PromiseQueueRoot {
+  add: <T>(action: () => Promise<T>) => Promise<T>;
+}
+
+interface PromiseQueue extends PromiseQueueRoot {
+  fork: () => PromiseQueue;
+  waitUntilIdle: () => Promise<undefined>;
+}
+
+function createGroup(root: PromiseQueueRoot): PromiseQueue {
+  let pendingCount = 0;
+  let idleDeferred: Deferred<undefined> | undefined;
+
+  return {
+    add: <T>(action: () => Promise<T>) => {
+      pendingCount++;
+      const finalize = <T>(result: T) => {
+        pendingCount--;
+        if (!pendingCount) {
+          idleDeferred?.resolve();
+          idleDeferred = undefined;
+        }
+        return result;
+      };
+      return root.add(() => action().then(finalize, finalize));
+    },
+    fork: () => createGroup(root),
+    waitUntilIdle: (): Promise<undefined> => {
+      if (!pendingCount) {
+        return Promise.resolve(undefined);
+      }
+
+      idleDeferred ??= createDeferred();
+      return idleDeferred.promise;
```

Comment:
I _think_ these semantics are right but this is another thing I would normally cover with unit tests (e.g. different combos of resolving before/between add+ waitUntilIdle)